core/src/main/scala/org/bitcoins/core/protocol/dlc/sign/DLCTxSigner.scala

```diff
@@ -225,10 +225,42 @@ case class DLCTxSigner(
     val cetsAndSigs = buildAndSignCETs(builder.contractInfo.allOutcomes)
     val (msgs, cets, sigs) = cetsAndSigs.unzip3
     val refundSig = signRefundTx
-
     (CETSignatures(msgs.zip(sigs), refundSig), cets)
   }
 
+  /** The equivalent of [[createCETsAndCETSigs()]] but async */
+  def createCETsAndCETSigsAsync()(implicit
+  ec: ExecutionContext): Future[(CETSignatures, Vector[WitnessTransaction])] = {
+    val outcomes = builder.contractInfo.allOutcomes
+    val outcomeBatchSize =
+      outcomes.length / Runtime.getRuntime.availableProcessors()
+    val batchSize =
+      Math.max(outcomeBatchSize, 1) //we have to at least have batchSize of 1
+
+    val cetsAndSigsF: Future[Vector[
+      Vector[(OracleOutcome, WitnessTransaction, ECAdaptorSignature)]]] = {
+      FutureUtil.batchAndParallelExecute[OracleOutcome,
+                                         Vector[(
+                                             OracleOutcome,
+                                             WitnessTransaction,
+                                             ECAdaptorSignature)]](
+        elements = outcomes,
+        f = outcomes =>
+          Future {
+            buildAndSignCETs(outcomes)
+          },
+        batchSize = batchSize)
+    }
+
+    val refundSig = signRefundTx
+
+    for {
+      cetsAndSigsNested <- cetsAndSigsF
+      cetsAndSigs = cetsAndSigsNested.flatten
+      (msgs, cets, sigs) = cetsAndSigs.unzip3
+    } yield (CETSignatures(msgs.zip(sigs), refundSig), cets)
```

Comment:
Might be worthing putting some `require`s here to make sure things are ordered how we expect them to be

Reply:
If you take a look at the implement of `FutureUtil.batchAndParallelExecute()` it preserves ordering: 

https://github.com/bitcoin-s/bitcoin-s/blob/5d6ba1235597a55adfbfcb29126e3a4dcbf7ebb3/core/src/main/scala/org/bitcoins/core/util/FutureUtil.scala#L108

assuming `f` preserves ordering. 

We seemed to preserve ordering in the sync version, but here is the code that gets called as `f` in the `batchAndParallelExecute()`

https://github.com/bitcoin-s/bitcoin-s/blob/5d6ba1235597a55adfbfcb29126e3a4dcbf7ebb3/core/src/main/scala/org/bitcoins/core/protocol/dlc/sign/DLCTxSigner.scala#L133